src/NonceTracker.ts

```diff
@@ -0,0 +1,251 @@
+ */
+export interface NonceDetails {
+  params: {
+    highestLocallyConfirmed: number;
+    nextNetworkNonce: number;
+    highestSuggested: number;
+  };
+  local: HighestContinuousFrom;
+  network: NetworkNextNonce;
+}
+
+/**
+ * @property nextNonce - The highest of the nonce values derived based on confirmed and pending transactions and eth_getTransactionCount method
+ * @property nonceDetails - details of nonce value derivation.
+ * @property releaseLock
+ */
+export interface NonceLock{
+  nextNonce: number;
+  nonceDetails: NonceDetails;
+  releaseLock: () => void;
+}
+
+/**
+ * @property name - The name for how the nonce was calculated based on the data used
+ * @property nonce - The next nonce value suggested by the eth_getTransactionCount method.
+ * @property blockNumber - The latest block from the network
+ * @property baseCount - Transaction count from the network suggested by eth_getTransactionCount method
+ */
+export interface NetworkNextNonce{
+  name: string;
+  nonce: number;
+  details: {
+    blockNumber: string;
+    baseCount: number;
+  };
+}
+
+/**
+ * @property name - The name for how the nonce was calculated based on the data used
+ * @property nonce - The next suggested nonce
+ * @property details{startPoint, highest} - the provided starting nonce that was used and highest derived from it (for debugging)
+ */
+export interface HighestContinuousFrom{
+  name: string;
+  nonce: number;
+  details: {
+    startPoint: number;
+    highest: number;
+  };
+}
+
+export interface Transaction {
+  status: string;
+  history: [Record<string, unknown>];
+  txParams: {
+    from: string;
+    gas: string;
+    value: string;
+    nonce: string;
+  };
+}
+
+export class NonceTracker {
+  private provider: Record<string, unknown>;
+
+  private blockTracker: any;
+
+  private ethQuery: any;
+
+  private getPendingTransactions: ((address: string) => Transaction[]);
+
+  private getConfirmedTransactions: ((address: string) => Transaction[]);
+
+  private lockMap: Record<string, unknown>;
+
+  constructor(opts: NonceTrackerOptions) {
+    this.provider = opts.provider;
+    this.blockTracker = opts.blockTracker;
+    this.ethQuery = new EthQuery(opts.provider);
+    this.getPendingTransactions = opts.getPendingTransactions;
+    this.getConfirmedTransactions = opts.getConfirmedTransactions;
+    this.lockMap = {};
+  }
+
+  /**
+   * @returns Promise<{ releaseLock: () => void }> with the key releaseLock (the global mutex)
+   */
+  async getGlobalLock(): Promise<{ releaseLock: () => void }> {
+    const globalMutex = this._lookupMutex('global');
+    // await global mutex free
+    const releaseLock = await globalMutex.acquire();
+    return { releaseLock };
+  }
+
+  /**
+   * this will return an object with the `nextNonce` `nonceDetails`, and the releaseLock
+   * Note: releaseLock must be called after adding a signed tx to pending transactions (or discarding).
+
```

Comment:
Formatting: empty line
```suggestion
   *
```